### ISSUE TITLE
ICE: Type aliases, nested types, and Fn cause panic

### ISSUE THREAD

**Issue by pierzchalski** [rust-lang/rust#25536]
A simple example involving a type alias, a nested type, and `Fn` causes an ICE.

I tried this code:
```rust
struct C<A> { a: A }

type T<A, B> = C<Fn(A) -> B>;

fn id<A, B>(t: T<&A, B>) -> T<&A, B> { t }
```

I expected to see either no error, or at most an error about lifetime elision.

Instead, when compiled with `rustc --crate-type=lib` I get an ICE. The effect persists when `C` is replaced by `Box`, but disappears if `Fn` is replaced by another two-parameter type.

The ICE:
```
error: internal compiler error: unexpected panic
thread 'rustc' panicked at 
'assertion failed: !predicates.has_escaping_regions()',
/build/rust/src/rustc-1.0.0/src/librustc_typeck/check/mod.rs:1698
```

The backtrace output is:
```
stack backtrace:
   1:     0x7f892d092189 - sys::backtrace::write::h85e36588f6ec519ed4r
   2:     0x7f892d099e96 - <unknown>
   3:     0x7f892d05d1a2 - rt::unwind::begin_unwind_inner::h58d8677c31c147e6t6v
   4:     0x7f892be6090e - <unknown>
   5:     0x7f892bead693 - <unknown>
   6:     0x7f892bee43d1 - check::wf::BoundsChecker<'cx, 'tcx>.TypeFolder<'tcx>::fold_ty::hf49f58e50ffe027105k
   7:     0x7f892bee619b - <unknown>
   8:     0x7f892bee46c8 - check::wf::BoundsChecker<'cx, 'tcx>.TypeFolder<'tcx>::fold_ty::hf49f58e50ffe027105k
   9:     0x7f892bedb8d7 - <unknown>
  10:     0x7f892bfc628d - <unknown>
  11:     0x7f892bfc16e0 - check_crate::h5570390160d0735dXcC
  12:     0x7f892d5cca18 - driver::phase_3_run_analysis_passes::h8472d32b1f7cb2d6nGa
  13:     0x7f892d5adb15 - driver::compile_input::hee759950e4b56686Qba
  14:     0x7f892d66efe1 - run_compiler::h57ce534eedce4c58z4b
  15:     0x7f892d66cc32 - <unknown>
  16:     0x7f892d66c169 - <unknown>
  17:     0x7f892d103b28 - <unknown>
  18:     0x7f892d103b15 - rust_try
  19:     0x7f892d66c418 - <unknown>
  20:     0x7f892d098d81 - <unknown>
  21:     0x7f8927b22373 - start_thread
  22:     0x7f892ccf227c - clone
  23:                0x0 - <unknown>
```

The output of `rustc --version --verbose` is:
```
rustc 1.0.0-beta.5 (built 2015-05-16)
binary: rustc
commit-hash: unknown
commit-date: unknown
build-date: 2015-05-16
host: x86_64-unknown-linux-gnu
release: 1.0.0-beta.5
```

**Comment by mitaa** [comment]
Simplified:

```rust
trait Foo<T> {}

fn bar<T>(t: Foo<&T>) -> Foo<&T> { t }
```

**Comment by leodasvacas** [comment]
Triage: According to [the playground](http://is.gd/lgXTAw), this no longer ICEs.

**Comment by arielb1** [comment]
duplicate of #26638
